File: yasmin_ros/src/yasmin_ros/ros_logs.cpp
// Copyright (C) 2024 Miguel Ángel González Santamarta
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <mutex>
#include <sstream>
#include <string>

#include <rclcpp/rclcpp.hpp>

#include "yasmin/logs.hpp"
#include "yasmin_ros/yasmin_node.hpp"

namespace yasmin_ros {

std::shared_ptr<rclcpp::Node> logger_node;
std::mutex logger_node_mutex;

void ros_log_message(yasmin::LogLevel level, const char *file,
                     const char *function, int line, const char *text) {

  std::shared_ptr<rclcpp::Node> local_node;
  {
    std::lock_guard<std::mutex> lock(logger_node_mutex);
    local_node = logger_node;
  }

  auto logger =
      local_node ? local_node->get_logger() : rclcpp::get_logger("yasmin_ros");

  auto format_msg = [&]() {
    std::ostringstream oss;
    oss << "[" << file << ":" << function << ":" << line << "] " << text;
    return oss.str();
  };

  switch (level) {
  case yasmin::LogLevel::ERROR:
    RCLCPP_ERROR(logger, "%s", format_msg().c_str());
    break;
  case yasmin::LogLevel::WARN:
    RCLCPP_WARN(logger, "%s", format_msg().c_str());
    break;
  case yasmin::LogLevel::INFO:
    RCLCPP_INFO(logger, "%s", format_msg().c_str());
    break;
  case yasmin::LogLevel::DEBUG:
    RCLCPP_DEBUG(logger, "%s", format_msg().c_str());
    break;
  }
}

void set_ros_loggers(rclcpp::Node::SharedPtr node) {

  {
    std::lock_guard<std::mutex> lock(logger_node_mutex);

    if (node == nullptr) {
      logger_node = YasminNode::get_instance();
    } else {
      logger_node = node;
    }
  }

  yasmin::set_loggers(ros_log_message);
}

} // namespace yasmin_ros

File: yasmin_ros/yasmin_ros/ros_logs.py
# Copyright (C) 2024 Miguel Ángel González Santamarta
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


from threading import Lock

import rclpy
from rclpy.node import Node

import yasmin
import yasmin_ros
from yasmin_ros.yasmin_node import YasminNode

__all__ = [
    "set_ros_loggers",
    "logger_node",
]

## Node used in the YASMIN ROS 2 logs
logger_node: Node = None
## Lock protecting logger_node access
_logger_node_lock: Lock = Lock()


def ros_log_message(
    level: yasmin.logs.LogLevel, file: str, function: str, line: int, text: str
) -> None:
    """
    Logs a message to the ROS 2 logger.

    This function logs a message using a ROS 2 logger.

    @param level The log level (ERROR, WARN, INFO, DEBUG).
    @param file The source file where the log function is called.
    @param function The function where the log function is called.
    @param line The line number in the source file.
    @param text The message to log.
    """

    message = f"[{file}:{function}:{line}] {text}"

    with _logger_node_lock:
        current_logger_node = yasmin_ros.logger_node

    logger = (
        current_logger_node.get_logger()
        if current_logger_node is not None
        else rclpy.logging.get_logger("yasmin_ros")
    )

    if level == yasmin.logs.LogLevel.ERROR:
        logger.error(message)

    elif level == yasmin.logs.LogLevel.WARN:
        logger.warning(message)

    elif level == yasmin.logs.LogLevel.INFO:
        logger.info(message)

    elif level == yasmin.logs.LogLevel.DEBUG:
        logger.debug(message)


def set_ros_loggers(node: Node = None) -> None:
    """
    Sets the ROS loggers for various logging levels.

    This function assigns the logging functions for info, warning,
    debug, and error messages to the Yasmin logging framework.

    @param node: ROS 2 node to use as logger node. If None, use YasminNode.
    @type node: Node
    """

    with _logger_node_lock:
        if node is None:
            yasmin_ros.logger_node = YasminNode.get_instance()
        else:
            yasmin_ros.logger_node = node

    yasmin.set_loggers(ros_log_message)
